File: Sean的leetcode/python/Serialize and Deserialize Binary Tree.py
###
# DFS pre-order
# Time Complexity: O(n)
# Space Complexity: O(h) = O(n)
###
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Codec:

    def serialize(self, root):
        """Encodes a tree to a single string.

        :type root: TreeNode
        :rtype: str
        """
        return self.dfs(root)

    def dfs(self, root):
        if not root:
            return '#'
        return str(root.val) + "," + self.dfs(root.left) + "," + self.dfs(root.right)

    def deserialize(self, data):
        """Decodes your encoded data to tree.

        :type data: str
        :rtype: TreeNode
        """
        return self.build(data.split(','))

    def build(self, values):
        v = values.pop(0)
        if v == '#':
            return None
        cur = TreeNode(int(v))
        cur.left = self.build(values)
        cur.right = self.build(values)
        return cur


# Your Codec object will be instantiated and called as such:
# codec = Codec()
# codec.deserialize(codec.serialize(root))
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Codec:

    def serialize(self, root):
        """Encodes a tree to a single string.
        
        :type root: TreeNode
        :rtype: str
        """
        re = []
        self.serial(root, re)
        return ",".join(re)
    
    def serial(self, root, re):
        if not root:
            re.append("#")
            return
        re.append(str(root.val))
        self.serial(root.left, re)
        self.serial(root.right, re)

    def deserialize(self, data):
        """Decodes your encoded data to tree.
        
        :type data: str
        :rtype: TreeNode
        """
        return self.deserial(iter(data.split(',')))
    
    def deserial(self, it):
        val = next(it)
        if val == "#":
            return None
        node = TreeNode(int(val))
        node.left = self.deserial(it)
        node.right = self.deserial(it)
        return node

# Your Codec object will be instantiated and called as such:
# codec = Codec()
# codec.deserialize(codec.serialize(root))

# Your Codec object will be instantiated and called as such:
# codec = Codec()
# codec.deserialize(codec.serialize(root))

File: MyLeetCode/Leetcode/Serialize and Deserialize Binary Tree.cpp
//https://leetcode.com/discuss/66147/recursive-preorder-python-and-c-o-n
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Codec {
public:

    void serial(TreeNode *root, string &res)
    {
        if (!root)
        {
            res += "#,";
            return;
        }
        res += to_string(root->val) + ",";
        serial(root->left, res);
        serial(root->right, res);
    }

    // Encodes a tree to a single string.
    string serialize(TreeNode* root) {
        string res;
        serial(root, res);
        return res;
    }

    TreeNode *deserial(string &data, int &pos)
    {
        if (pos == data.size())
            return 0;
        int end = data.find_first_of(',', pos);
        string s = data.substr(pos, end - pos);
        pos = end + 1;
        if (s == "#")
            return 0;
        TreeNode *root = new TreeNode(stoi(s));
        root->left = deserial(data, pos);
        root->right = deserial(data, pos);
        return root;
    }

    // Decodes your encoded data to tree.
    TreeNode* deserialize(string data) {
        int pos = 0;
        return deserial(data, pos);
    }
};

// Your Codec object will be instantiated and called as such:
// Codec codec;
// codec.deserialize(codec.serialize(root));